src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+
+            updatedPeers + (healthProposal.checkingPeerId -> prd)
+          }.getOrElse(roundPeers)
+        (updated, updated)
+      }
+    } yield ()
+
+  private def fetchProposals(): F[Unit] =
+    for {
+      isAcquired <- fetchingProposalLock.tryAcquire
+      _ <- if (isAcquired)
+        fetchHealthProposals()
+          .handleErrorWith(
+            e =>
+              fetchingProposalLock.release >> logger
+                .error(e)("Error fetching healthcheck proposal.") /*>> e.raiseError[F, Unit]*/
+          ) //not needed I think because every fetch failure is handled in fetchHealthProposals, should do the same for sendHealthProposal I think
+      else
+        F.unit
+      _ <- fetchingProposalLock.release
+    } yield ()
+
+  private def fetchHealthProposals(): F[Unit] =
+    for {
+      peers <- roundPeers.get
+        .map(_.filter { case (_, rd) => rd.consensusData.isEmpty })
+      roundIds <- roundIds.get
+      responses <- peers.toList.traverse {
+        case (id, roundData) => {
+          {
+            roundData.peerData match {
+              case Left(proxyPeers) =>
+                for {
+                  proxyPeer <- randomlyPickProxyPeer(proxyPeers.peersData)
+                  response <- PeerResponse.run(
+                    apiClient.healthcheck.fetchPeerHealthStatus(FetchPeerHealthStatus(roundIds.toSortedSet, id.some)),
+                    unboundedBlocker
+                  )(proxyPeer.peerMetadata.toPeerClientMetadata.copy(port = "9003"))
+                } yield response
+              case Right(peerData) =>
+                PeerResponse.run(
+                  apiClient.healthcheck.fetchPeerHealthStatus(FetchPeerHealthStatus(roundIds.toSortedSet)),
+                  unboundedBlocker
+                )(peerData.peerMetadata.toPeerClientMetadata.copy(port = "9003"))
+            }
+          }.handleErrorWith(
+            _ => FetchingFailed(id).asInstanceOf[FetchProposalError].asLeft[SendPerceivedHealthStatus].pure[F]
+          )
+        }.map(_.bimap(id -> _, id -> _))
+      }
+      (failed, succeeded) = responses.separate
+      _ <- succeeded.traverse { case (_, proposal) => processProposal(proposal.consensusHealthStatus) }
+      _ <- failed.traverse {
+        case (id, error) =>
+          logger.debug(error)(s"Error during fetching healtcheck proposal from peer with id=$id.")
+      }
+    } yield ()
+
+  private def randomlyPickProxyPeer(proxyPeers: NonEmptySet[PeerData]): F[PeerData] = F.delay(
+    Random.shuffle(proxyPeers.toList).head
+  )
+
+  private def sendProposal(): F[Unit] =
+    for {
+      isAcquired <- sendingProposalLock.tryAcquire
+      _ <- if (isAcquired)
+        sendHealthStatus()
+          .handleErrorWith(
+            e =>
+              sendingProposalLock.release >> logger
+                .error(e)("Error sending healtcheck proposal") /*>> e.raiseError[F, Unit]*/
+          )
+      else
+        F.unit
+      _ <- sendingProposalLock.release
```

Comment:
why these checks for acquiring are needed? can't you just acquire the semaphore, do what you have to do and release it then? if another process will be in progress, it will wait for acquiring the semaphore. currently, looks like if anything is in progress, it forcibly releases it anyway.

Reply:
It runs periodically so if another process is in progress it should not start. I probably made a mistake of the last command running every time -> so releasing even if not acquired, it should be in the first if also for the positive case.